src/sparse_blas/backends/cusparse/cusparse_error.hpp

```diff
@@ -0,0 +1,100 @@
+
+#include <string>
+
+#include <cuda.h>
+#include <cusparse.h>
+
+#include "oneapi/mkl/exceptions.hpp"
+
+namespace oneapi::mkl::sparse::cusparse {
+
+inline std::string cuda_result_to_str(CUresult result) {
+    switch (result) {
+#define ONEMKL_CUSPARSE_CASE(STATUS) \
+    case STATUS: return #STATUS
+        ONEMKL_CUSPARSE_CASE(CUDA_SUCCESS);
+        ONEMKL_CUSPARSE_CASE(CUDA_ERROR_NOT_PERMITTED);
+        ONEMKL_CUSPARSE_CASE(CUDA_ERROR_INVALID_CONTEXT);
+        ONEMKL_CUSPARSE_CASE(CUDA_ERROR_INVALID_DEVICE);
+        ONEMKL_CUSPARSE_CASE(CUDA_ERROR_INVALID_VALUE);
+        ONEMKL_CUSPARSE_CASE(CUDA_ERROR_OUT_OF_MEMORY);
+        ONEMKL_CUSPARSE_CASE(CUDA_ERROR_LAUNCH_OUT_OF_RESOURCES);
+        default: return "<unknown>";
+    }
+}
+
+#define CUDA_ERROR_FUNC(func, ...)                                                  \
+    do {                                                                            \
+        auto res = func(__VA_ARGS__);                                               \
+        if (res != CUDA_SUCCESS) {                                                  \
+            throw oneapi::mkl::exception("sparse_blas", #func,                      \
+                                         "cuda error: " + cuda_result_to_str(res)); \
+        }                                                                           \
+    } while (0)
+
+inline std::string cusparse_status_to_str(cusparseStatus_t status) {
+    switch (status) {
+#define ONEMKL_CUSPARSE_CASE(STATUS) \
+    case STATUS: return #STATUS
+        ONEMKL_CUSPARSE_CASE(CUSPARSE_STATUS_SUCCESS);
+        ONEMKL_CUSPARSE_CASE(CUSPARSE_STATUS_NOT_INITIALIZED);
+        ONEMKL_CUSPARSE_CASE(CUSPARSE_STATUS_ALLOC_FAILED);
+        ONEMKL_CUSPARSE_CASE(CUSPARSE_STATUS_INVALID_VALUE);
+        ONEMKL_CUSPARSE_CASE(CUSPARSE_STATUS_ARCH_MISMATCH);
+        ONEMKL_CUSPARSE_CASE(CUSPARSE_STATUS_EXECUTION_FAILED);
+        ONEMKL_CUSPARSE_CASE(CUSPARSE_STATUS_INTERNAL_ERROR);
+        ONEMKL_CUSPARSE_CASE(CUSPARSE_STATUS_MATRIX_TYPE_NOT_SUPPORTED);
+        ONEMKL_CUSPARSE_CASE(CUSPARSE_STATUS_NOT_SUPPORTED);
+        ONEMKL_CUSPARSE_CASE(CUSPARSE_STATUS_INSUFFICIENT_RESOURCES);
+#undef ONEMKL_CUSPARSE_CASE
+        default: return "<unknown>";
+    }
+}
+
+inline void check_status(cusparseStatus_t status, const std::string& function,
+                         std::string error_str = "") {
+    if (status != CUSPARSE_STATUS_SUCCESS) {
+        if (!error_str.empty()) {
+            error_str += "; ";
+        }
+        error_str += "cuSPARSE status: " + cusparse_status_to_str(status);
+        switch (status) {
+            case CUSPARSE_STATUS_NOT_SUPPORTED:
+                throw oneapi::mkl::unimplemented("sparse_blas", function, error_str);
+            case CUSPARSE_STATUS_INVALID_VALUE:
+                throw oneapi::mkl::invalid_argument("sparse_blas", function, error_str);
+            default: throw oneapi::mkl::exception("sparse_blas", function, error_str);
```

Comment:
Why are we not handling other statuses here?

Reply:
I couldn't find good equivalent exceptions for the other cuSPARSE statuses so I just throw a generic exception. Not sure what else I could do here.

Reply:
So the mapping based on https://oneapi-spec.uxlfoundation.org/specifications/oneapi/latest/elements/onemkl/source/architecture/architecture#exception-classification would be something like:
|cuSPARSE error|oneMKL exception|
|-----------------|--------------------|
|CUSPARSE_STATUS_NOT_INITIALIZED|uninitialized|
|CUSPARSE_STATUS_ALLOC_FAILED|device_bad_alloc|
|CUSPARSE_STATUS_INVALID_VALUE|invalid_argument|
|CUSPARSE_STATUS_NOT_SUPPORTED|unimplemented|
|CUSPARSE_STATUS_ARCH_MISMATCH|unsupported_device|
|CUSPARSE_STATUS_EXECUTION_FAILED|computation_error|
|CUSPARSE_STATUS_INTERNAL_ERROR|computation_error|
|CUSPARSE_STATUS_MATRIX_TYPE_NOT_SUPPORTED|unimplemented|
|CUSPARSE_STATUS_NOT_SUPPORTED|unimplemented|
|CUSPARSE_STATUS_INSUFFICIENT_RESOURCES|generic exception text|

Reply:
I don't fully agree with this table:

- `CUSPARSE_STATUS_ALLOC_FAILED` can be either a device side or host side allocation failure. Using `device_bad_alloc` could be misleading.
- `CUSPARSE_STATUS_ARCH_MISMATCH` could make sense but the `unsupported_device` exception isn't easy to use, it requires the SYCL device. It would be a large change to capture the SYCL device for every occurrences of `check_status` or `CUSPARSE_ERR_FUNC` (54 occurrences in total). I think that's the reason why most backends only use this exception "early" in the oneMKL stack frame (i.e. when dispatching to a backend).
- `CUSPARSE_STATUS_EXECUTION_FAILED` and `CUSPARSE_STATUS_INTERNAL_ERROR` are very generic errors in cuSPARSE. Reducing them to a `computation_error` is misleading.
- `CUSPARSE_STATUS_MATRIX_TYPE_NOT_SUPPORTED` is confusing to me. The name sound like it could match with an `unimplemented` exception but its description sound more like an invalid value was set or a field was not initialized. We need to be careful with `unimplemented` as it is used to skip the tests, I've thrown an `invalid_argument` instead.

I've caught 2 more statuses in https://github.com/oneapi-src/oneMKL/pull/527/commits/9c164252e29049d4fd632b72f374741738144084. The rest of the statuses fallback to the generic exception. I don't expect users need more detailed exception types, the message of the exception is more useful as it contains the original backend status.

Reply:
On the Intel oneMKL sparse BLAS side, we've mostly used `computation_error` for generic internal errors, but I see your point. What you have now is fine then since we also throw back the cuSPARSE status in the exception string.